presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -452,13 +462,63 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
         }
 
         HiveDirectoryContext hiveDirectoryContext = new HiveDirectoryContext(recursiveDirWalkerEnabled ? RECURSE : IGNORED, cacheable);
+
+        if (HiveSessionProperties.isListFilesDisabled(session) && partition.isPresent()) {
+            Map<String, String> parameters = partition.get().getParameters();
+            if (parameters.containsKey(FILE_NAMES) && parameters.containsKey(FILE_SIZES)) {
+                List<String> fileNames = getFileNames(parameters.get(FILE_NAMES));
+                List<Long> fileSizes = Arrays.stream(parameters.get(FILE_SIZES).split(",")).map(Long::valueOf).collect(toImmutableList());
+
+                // Verify that the count of fileNames and fileSizes are same
+                verify(fileNames.size() == fileSizes.size(), "List of fileNames and fileSizes differ in length");
+
+                ImmutableList.Builder<HiveFileInfo> fileListBuilder = ImmutableList.builder();
+                for (int i = 0; i < fileNames.size(); i++) {
+                    Path filePath = new Path(path, fileNames.get(i));
+                    FileStatus fileStatus = new FileStatus(fileSizes.get(i), false, 1, TARGET_BLOCK_SIZE.toBytes(), 0, filePath);
+                    try {
+                        BlockLocation[] locations = fileSystem.getFileBlockLocations(fileStatus, 0, fileStatus.getLen());
```

Comment:
This will still call the underlying remote FS. I think `locations` only works for colocated HDFS. It should be empty.